interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,966 @@
+}
+
+// addRulePermissionToTree adds all the path pattern variants for the given
+// rule to the map for the given permission.
+//
+// Returns whether the DB needs to be saved, either because the rule was
+// successfully added or because an expired rule was removed.
+//
+// If there are conflicting pattern variants from other non-expired rules,
+// all variants which were previously added during this function call are
+// removed from the path map, and an error is returned along with a list of
+// the conflicting variants and the IDs of the conflicting rules.
+//
+// Conflicts with expired rules, however, result in the expired rule being
+// immediately removed, and the new rule can continue to be added.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) (needToSave bool, err error, conflicts []RuleConflict) {
+	permVariants := rdb.permissionDBForUserSnapInterfacePermission(rule.User, rule.Snap, rule.Interface, permission)
+
+	newVariantEntries := make(map[string]variantEntry, rule.Constraints.PathPattern.NumVariants())
+	addVariant := func(index int, variant patterns.PatternVariant) {
+		newEntry := variantEntry{
+			Variant: variant,
+			RuleID:  rule.ID,
+		}
+		variantStr := variant.String()
+		conflictingVariantEntry, exists := permVariants.VariantEntries[variantStr]
+		switch {
+		case !exists:
+			newVariantEntries[variantStr] = newEntry
+		case rdb.removeIfExpired(conflictingVariantEntry.RuleID, rule.Timestamp):
+			needToSave = true // removed expired rule, or was inconsistency
+			newVariantEntries[variantStr] = newEntry
+		default:
+			// Exists and is not expired, so there's a conflict
+			conflicts = append(conflicts, RuleConflict{
+				Variant:       variantStr,
+				ConflictingID: conflictingVariantEntry.RuleID,
+			})
+		}
+	}
+	rule.Constraints.PathPattern.RenderAllVariants(addVariant)
+
+	if len(conflicts) > 0 {
+		return needToSave, ErrPathPatternConflict, conflicts
+	}
+
+	for variantStr, entry := range newVariantEntries {
+		permVariants.VariantEntries[variantStr] = entry
+	}
+
+	return true, nil, nil
+}
+
+// removeIfExpired removes the rule with the given ID from the database if it
+// is expired. Returns true if the rule was removed, couldn't be found, or had
+// some sort of internal inconsistency.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) removeIfExpired(id prompting.IDType, currTime time.Time) bool {
+	rule, err := rdb.ruleWithID(id)
+	if err != nil {
+		return true
+	}
+	expired, err := rule.Expired(currTime)
+	switch {
+	case err != nil:
+		// Issue with expiration, this should not occur
+		logger.Noticef("error while checking whether rule had expired: %v", err)
+		fallthrough
+	case expired:
+		// Remove expired conflicting rule from DB
+		rdb.removeRuleFromTree(rule) // if error occurs, rule still fully removed
+		rdb.removeRuleWithID(rule.ID)
```

Comment:
we do this two things together in a lot of places, maybe we should have a removeRule(rule) helper that does both?